From 20ae4089d0afeb24e9ceb026b996bfa55c983cc2 Mon Sep 17 00:00:00 2001
From: Abhishek Pandit-Subedi <abhishekpandit@chromium.org>
Date: Fri, 11 Sep 2020 15:33:18 -0700
Subject: [PATCH] Bluetooth: Only mark socket zapped after unlocking

Since l2cap_sock_teardown_cb does not acquire the channel lock before
setting the socket as zapped, it can race with l2cap_sock_release, which
frees the socket.  Mark the socket as zapped only after the socket cleanup
has completed.

This is upstream commit 20ae4089d0afeb24e9ceb026b996bfa55c983cc2.
It was backported to the 5.4.y stable series.
---
 net/bluetooth/l2cap_sock.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/net/bluetooth/l2cap_sock.c b/net/bluetooth/l2cap_sock.c
--- a/net/bluetooth/l2cap_sock.c
+++ b/net/bluetooth/l2cap_sock.c
@@ -1263,8 +1263,6 @@ static void l2cap_sock_teardown_cb(struct l2cap_chan *chan, int err)
 
 	parent = bt_sk(sk)->parent;
 
-	sock_set_flag(sk, SOCK_ZAPPED);
-
 	switch (chan->state) {
 	case BT_OPEN:
 	case BT_BOUND:
@@ -1291,5 +1289,8 @@ static void l2cap_sock_teardown_cb(struct l2cap_chan *chan, int err)
 		break;
 	}
-
 	release_sock(sk);
+
+	/* Only zap after cleanup to avoid use after free race */
+	sock_set_flag(sk, SOCK_ZAPPED);
+
 }
-- 
2.25.1
